fedimint-api-client/src/api/global_api/with_online_toggle.rs

```diff
@@ -0,0 +1,95 @@
+use std::collections::BTreeSet;
+
+use anyhow::format_err;
+use fedimint_core::core::ModuleInstanceId;
+use fedimint_core::module::ApiRequestErased;
+use fedimint_core::task::{MaybeSend, MaybeSync};
+use fedimint_core::{apply, async_trait_maybe_send, PeerId};
+use serde_json::Value;
+use tokio::sync::{watch, RwLock};
+
+use super::super::{DynModuleApi, IRawFederationApi};
+use crate::api::{PeerError, PeerResult};
+
+/// Convenience extension trait used for wrapping [`IRawFederationApi`] in
+/// a [`RawFederationApiWithOnlineToggle`]
+pub trait RawFederationApiWithOnlineToggleExt
+where
+    Self: Sized,
+{
+    fn with_online_toggle(
+        self,
+        online: watch::Receiver<bool>,
+    ) -> RawFederationApiWithOnlineToggle<Self>;
+}
+
+impl<T> RawFederationApiWithOnlineToggleExt for T
+where
+    T: IRawFederationApi + MaybeSend + MaybeSync + 'static,
+{
+    fn with_online_toggle(
+        self,
+        online: watch::Receiver<bool>,
+    ) -> RawFederationApiWithOnlineToggle<T> {
+        RawFederationApiWithOnlineToggle::new(self, online)
+    }
+}
+
+/// [`IRawFederationApi`] wrapping some `T: IRawFederationApi` and adding
+/// a toggle for online/offline state.
+///
+/// If the state is offline, all requests will block until toggle switched to
+/// online, without actually making any connections.
+///
+/// Use [`RawFederationApiWithOnlineToggleExt::with_online_toggle`] to
+/// create.
+#[derive(Debug)]
+pub struct RawFederationApiWithOnlineToggle<T> {
+    pub(crate) inner: T,
+    online: tokio::sync::RwLock<watch::Receiver<bool>>,
```

Comment:
why a watch channel?  instead of say `Arc<AtomicBool>`

Reply:
So it's possible to block on it waiting for updates (which currently I'm not using, but  I started with, then reverted, then wondered if I should do something in between - see comment in the code).